### DISCUSSION TITLE
Setting EdgeX MessageBus to use MQTT Pub/Sub

### DISCUSSION THREAD

**Discussion by neerajasjawali** [edgexfoundry/community#56] [Q&A]
Hi all,
I am working on setting the MessageBus configuration for EdgeX (jakarta) to use MQTT Pub/Sub.
I have modified the configuration.toml file of core-data, app-service and device-service. Moqsuitto broker is running, While running the device-service for reading camera data, I am getting following errors on various services terminal. 

Error on core-data service:
`level=ERROR ts=2023-02-16T04:55:59.428261237Z app=core-data source=subscriber.go:72 msg="fail to persist the event, Event Id exists"`

Error on metadata service:
`level=INFO ts=2023-02-16T06:23:47.431125067Z app=core-metadata source=message.go:58 msg="Service started in: 31.079106ms"
level=ERROR ts=2023-02-16T05:33:47.746644181Z app=device-virtual source=utils.go:71 msg="Failed to publish event to MessageBus: Timeout occured while performing a 'Publish' operation: Unable to publish message"`

Error on device-service:
`level=ERROR ts=2023-02-16T05:33:47.746644181Z app=device-virtual source=utils.go:71 msg="Failed to publish event to MessageBus: Timeout occured while performing a 'Publish' operation: Unable to publish message"`

Error on mosquitto broker:
```
New connection from 10.168.1.190 on port 1883.
1676525380: New client connected from 10.168.1.190 as core-data (p2, c1, k10).
1676525590: Client core-data has exceeded timeout, disconnecting.
1676525605: New connection from 10.168.1.190 on port 1883.
1676525605: New client connected from 10.168.1.190 as core-data (p2, c1, k10).
1676525629: Client mqtt-app-service has exceeded timeout, disconnecting.
```

Attached the screenshots of configuration.toml files
![Screenshot from 2023-02-16 12-33-55](https://user-images.githubusercontent.com/50900251/219292720-db6a94ce-749f-4676-9ee3-135b62f11338.png)

![Screenshot from 2023-02-16 12-34-47](https://user-images.githubusercontent.com/50900251/219292735-bfe8cf69-c1e7-4c31-9ae5-30d528e68726.png)

It would be helpful, if you could give me suggestions to resolve the issue.
I would also like to know the modifications to be done in mqtt broker configuration. 

Thanks in advance

**Comment by herrdragon** [comment]
Hi, have you looked at this https://docs.edgexfoundry.org/2.1/microservices/application/Triggers/#external-mqtt-trigger 
Try using :
[Trigger]
Type="external-mqtt" 
[Trigger.ExternalMqtt]
    Url = "tcp://edgex-mqtt-broker:1883" #  fully qualified URL to connect to the MQTT broker
    SubscribeTopics="SomeTopics"...

The way you have is more for redis pub-sub

**Reply by neerajasjawali** [reply]
Hi @herrdragon 
The configuration is likely similar to redis pub/sub. I looked at this https://docs.edgexfoundry.org/2.1/microservices/application/Triggers/. Within which they have given an example for using mqtt, I referred the same and made the changes. As you can see I have written the type as mqtt and modified some changes as below
```
[Writable.Pipeline.Functions.MQTTExport]                            
      [Writable.Pipeline.Functions.MQTTExport.Parameters]                  
      BrokerAddress = "tcp://edgex-mqtt-broker:1883" # Hive Public Test Broker
      Topic = "edgex-export"                            
      SecretPath = "mqtt"                             
      ClientId = "MQTT-Export"                      
      QOS="0"                                                           
      AutoReconnect="true"                                                 
      KeepAlive = "10m" # Empty indicates to use default value                
      ConnectTimeout = "5m" # Empty indicates to use default value
      Retain="true"                                                        
      SkipVerify = "false"                          
      AuthMode = "none"                                                 
      PersistOnError = "false"  
```
should I not use mqtt-export and instead use external-mqtt?
If Yes, then why?
I would also like to know if we are using mosquitto broker for pub sub, do I need to modify the mosquitto configuration file and how?
Thanks in advance

**Reply by herrdragon** [reply]
I can’t tell you in details why, all I know if that I, like you, am using mosquito as a broker and since it is not part of edgex, I use the external-mqtt configuration for the [Trigger] part and then [Trigger.ExternalMqtt] for the broker config and is working fine with no issues. Maybe someone from edgex here can explain why is not working the way you have. Hope this helps you.

**Reply by neerajasjawali** [reply]
Hi @herrdragon,

Thank you for your suggestion.

**Reply by lenny-goodell** [reply]
FYI, `external-mqtt` is only for App Service to receive "trigger" data from an external MQTT broker instead of the internal MessageBus. 
It is not applicable to switching the internal MessageBus to use MQTT.

**Reply by neerajasjawali** [reply]
Hi @lenny-intel 
Thank you for explaining the difference.

**Comment by lenny-goodell** [comment]
@neerajasjawali , are you running in Docker?
If so, the easiest way to switch to using MQTT for the internal MessageBus is to use compose builder to generate a compose file properly configured for this.
https://github.com/edgexfoundry/edgex-compose/blob/jakarta/compose-builder/README.md#gen
The base command is `make gen no-secty mqtt-bus <device-service>|<app-service>`
See README linked above for all options.

**Reply by neerajasjawali** [reply]
Hi @lenny-intel 
I am not running in Docker. I modified the mqtt-export configuration as explained in this link [https://github.com/edgexfoundry/edgex-compose/blob/jakarta/compose-builder/add-mqtt-messagebus-app-template.yml].

**Reply by lenny-goodell** [reply]
@neerajasjawali , the problem is your host name. It should just be `localhost`. The one you have is fore use in Docker. So change all instances of `edgex-mqtt-broker` to `localhost`

**Reply by neerajasjawali** [reply]
@lenny-intel, earlier I had defined the host name as localhost instead of edgex-mqtt-broker but it was throwing the below error. 
```
root@nb2:~/secret_jakarta/edgex_secretstore/edgex-go/cmd/security-secretstore-setup# ./security-secretstore-setup --vaultInterval=10
level=INFO ts=2023-01-18T12:38:15.631135754Z app=security-secretstore-setup source=config.go:373 msg="Loaded service configuration from ./res/configuration.toml"
level=INFO ts=2023-01-18T12:38:15.642611306Z app=security-secretstore-setup source=variables.go:352 msg="Variables override of 'SecretStore.Port' by environment variable: SECRETSTORE_PORT=8200"
level=INFO ts=2023-01-18T12:38:15.643620179Z app=security-secretstore-setup source=variables.go:352 msg="Variables override of 'SecretStore.Host' by environment variable: SECRETSTORE_HOST=edgex-vault"
level=INFO ts=2023-01-18T12:38:15.64511599Z app=security-secretstore-setup source=config.go:510 msg="Using local configuration from file (2 envVars overrides applied)"
level=INFO ts=2023-01-18T12:38:15.645940886Z app=security-secretstore-setup source=init.go:101 msg="bypassing certificate verification for secret store connection"
level=INFO ts=2023-01-18T12:38:15.646688792Z app=security-secretstore-setup source=init.go:118 msg="SecretStoreClient created"
level=INFO ts=2023-01-18T12:38:15.647149734Z app=security-secretstore-setup source=init.go:134 msg="vault master key encryption not enabled. IKM_HOOK not set."
level=ERROR ts=2023-01-18T12:38:15.653840889Z app=security-secretstore-setup source=request.go:78 msg="unable to make request to health check failed: Get \"http://edgex-vault:8200/v1/sys/health\": dial tcp 10.168.1.0:8200: connect: connection refused"
level=ERROR ts=2023-01-18T12:38:15.654289833Z app=security-secretstore-setup source=init.go:217 msg="vault is in an unknown state. No Status code available"
level=INFO ts=2023-01-18T12:38:15.654585795Z app=security-secretstore-setup source=init.go:231 msg="trying Vault init/unseal again in 10 seconds"
level=ERROR ts=2023-01-18T12:38:25.656324655Z app=security-secretstore-setup source=request.go:78 msg="unable to make request to health check failed: Get \"http://edgex-vault:8200/v1/sys/health\": dial tcp 10.168.1.0:8200: connect: connection refused"
level=ERROR ts=2023-01-18T12:38:25.656536628Z app=security-secretstore-setup source=init.go:217 msg="vault is in an unknown state. No Status code available"
level=INFO ts=2023-01-18T12:38:25.656702607Z app=security-secretstore-setup source=init.go:231 msg="trying Vault init/unseal again in 10 seconds"
level=ERROR ts=2023-01-18T12:38:35.666435458Z app=security-secretstore-setup source=request.go:78 msg="unable to make request to health check failed: Get \"http://edgex-vault:8200/v1/sys/health\": dial tcp 10.168.1.0:8200: connect: connection refused"
level=ERROR ts=2023-01-18T12:38:35.666628434Z app=security-secretstore-setup source=init.go:217 msg="vault is in an unknown state. No Status code available"
level=INFO ts=2023-01-18T12:38:35.666770416Z app=security-secretstore-setup source=init.go:231 msg="trying Vault init/unseal again in 10 seconds"
level=ERROR ts=2023-01-18T12:38:45.680154806Z app=security-secretstore-setup source=request.go:78 msg="unable to make request to health check failed: Get \"http://edgex-vault:8200/v1/sys/health\": dial tcp 10.168.1.0:8200: connect: connection refused"
level=ERROR ts=2023-01-18T12:38:45.680751731Z app=security-secretstore-setup source=init.go:217 msg="vault is in an unknown state. No Status code available"
level=INFO ts=2023-01-18T12:38:45.680991701Z app=security-secretstore-setup source=init.go:231 msg="trying Vault init/unseal again in 10 seconds"
level=ERROR ts=2023-01-18T12:38:55.682841547Z app=security-secretstore-setup source=request.go:78 msg="unable to make request to health check failed: Get \"http://edgex-vault:8200/v1/sys/health\": dial tcp 10.168.1.0:8200: connect: connection refused"
level=ERROR ts=2023-01-18T12:38:55.683088516Z app=security-secretstore-setup source=init.go:217 msg="vault is in an unknown state. No Status code available"
level=INFO ts=2023-01-18T12:38:55.683254495Z app=security-secretstore-setup source=init.go:231 msg="trying Vault init/unseal again in 10 seconds"
```
The reason was because security-secretestore thinks localhost as 127.0.0.1, but whereas vault considers localhost as my system ip address . And in the edgex ted talk they mentioned that giving the host name as local host adds some restrictions, like we can not check the status of services in API.

**Reply by lenny-goodell** [reply]
This has nothing to do with MQTT.
How are you running `security-secretstore-setup`, Docker or non-Docker? 
The name `edgex-vault` in URLs above is only valid running in Docker.
All this is so much easier when you run all the services in Docker.

**Reply by neerajasjawali** [reply]
@lenny-intel, I am running services without Docker

**Reply by lenny-goodell** [reply]
> I am running services without Docker

Why?
There is a reason we use Docker....

**Reply by bnevis-i** [reply]
You need to do some hostname debugging.  When vault is started, what does the "listener" configuration block say to use as the address (the docker version uses edgex-vault:8200, the snap version uses localhost:8200).  Also what is your host ip address (ip addr list).  What does localhost and edgex-vault resolve to (nslookup localhost, nslookup edgex-vault, nslookup $(hostname))

Listening on hostname or 0.0.0.0 means that it is reachable from a remote machine on the network.

If vault isn't listening on all interfaces (0.0.0.0), then the clients need to be connecting to vault on the same interface that it is listening on (either localhost or edgex-vault).  If you're not running docker then I wouldn't recommend using edgex-vault, because any 'ol system on your network can claim that name and suddenly all your vault traffic is going to who-knows-where.

**Reply by neerajasjawali** [reply]
> > I am running services without Docker
> 
> Why? There is a reason we use Docker....

I am assigned to test EdgeX security without Docker and other teammate is looking into running with Docker.

**Reply by neerajasjawali** [reply]
> You need to do some hostname debugging. When vault is started, what does the "listener" configuration block say to use as the address (the docker version uses edgex-vault:8200, the snap version uses localhost:8200). Also what is your host ip address (ip addr list). What does localhost and edgex-vault resolve to (nslookup localhost, nslookup edgex-vault, nslookup $(hostname))
> 
> Listening on hostname or 0.0.0.0 means that it is reachable from a remote machine on the network.
> 
> If vault isn't listening on all interfaces (0.0.0.0), then the clients need to be connecting to vault on the same interface that it is listening on (either localhost or edgex-vault). If you're not running docker then I wouldn't recommend using edgex-vault, because any 'ol system on your network can claim that name and suddenly all your vault traffic is going to who-knows-where.


Vault listener block:

 ```
listener "tcp" { 
              address = "edgex-vault:8200" 
              tls_disable = "1" 
              cluster_address = "edgex-vault:8201" 
          } 
          backend "file" {
              path = "/home/root/secret_jakarta/edgex_secretstore/vault/file"
          } 
          default_lease_ttl = "168h" 
          max_lease_ttl = "720h"

```
Within the ip address list we have defined edgex-vault,edgex-core-consul,edgex-redis and so on with the system ip address (10.168.1.0)
